File: nodejs-soa/order-service/src/application/OrderService.ts
// =============================================================================
// Order Service Implementation - Saga Orchestrator
// =============================================================================
// This service implements the Saga pattern for distributed transactions.
//
// Saga Flow:
// 1. Client calls create() → Order saved as PENDING
// 2. order.created event published → Inventory Service processes it
// 3. Inventory responds with stock.reserved or stock.insufficient
// 4. Order Service updates order status based on response
//
// Compensating Transaction:
// - If order is cancelled after confirmation → order.cancelled event published
// - Inventory Service releases the reserved stock
// =============================================================================

import { Order, CreateOrderData } from '../domain/model/Order';
import { OrderStatus, isValidTransition } from '../domain/model/OrderStatus';
import { IOrderService } from '../domain/service/IOrderService';
import { OrderRepository } from '../infrastructure/persistence/OrderRepository';
import { KafkaProducer } from '../infrastructure/messaging/KafkaProducer';
import { createOrderCreatedEvent, createOrderConfirmedEvent, createOrderCancelledEvent } from '../infrastructure/messaging/events';

export class OrderService implements IOrderService {
  constructor(
    private readonly repository: OrderRepository,
    private readonly kafkaProducer: KafkaProducer
  ) {}

  async findAll(): Promise<Order[]> {
    return this.repository.findAll();
  }

  async findById(id: string): Promise<Order | null> {
    return this.repository.findById(id);
  }

  async findByUserId(userId: string): Promise<Order[]> {
    return this.repository.findByUserId(userId);
  }

  /**
   * Create a new order - Step 1 of the Saga.
   * The order starts in PENDING state until inventory confirms stock availability.
   */
  async create(data: CreateOrderData): Promise<Order> {
    // 1. Persist the order with PENDING status
    const order = await this.repository.create(data);

    // 2. Publish order.created event to trigger inventory check
    try {
      const event = createOrderCreatedEvent(order);
      await this.kafkaProducer.publish('order.created', order.id, event);
      console.log(`[OrderService] Saga started - Published order.created for order: ${order.id}`);
    } catch (error) {
      console.error(`[OrderService] Failed to publish order.created event:`, error);
      // In production, use outbox pattern to guarantee event delivery
    }

    return order;
  }

  /**
   * Confirm an order - Called when Inventory Service reserves stock.
   * This is Step 3 of the Saga (happy path).
   */
  async confirm(id: string): Promise<Order | null> {
    const order = await this.repository.findById(id);
    if (!order) return null;

    if (!isValidTransition(order.status, OrderStatus.CONFIRMED)) {
      console.warn(`[OrderService] Invalid transition: ${order.status} → CONFIRMED for order ${id}`);
      return order;
    }

    const updatedOrder = await this.repository.updateStatus(id, OrderStatus.CONFIRMED);

    // Publish order.confirmed event
    if (updatedOrder) {
      try {
        const event = createOrderConfirmedEvent(updatedOrder);
        await this.kafkaProducer.publish('order.confirmed', id, event);
        console.log(`[OrderService] Saga completed - Order ${id} confirmed`);
      } catch (error) {
        console.error(`[OrderService] Failed to publish order.confirmed event:`, error);
      }
    }

    return updatedOrder;
  }

  /**
   * Cancel an order - Triggers compensating transaction.
   * Publishes order.cancelled so Inventory Service can release reserved stock.
   */
  async cancel(id: string): Promise<Order | null> {
    const order = await this.repository.findById(id);
    if (!order) return null;

    if (!isValidTransition(order.status, OrderStatus.CANCELLED)) {
      console.warn(`[OrderService] Invalid transition: ${order.status} → CANCELLED for order ${id}`);
      return order;
    }

    const updatedOrder = await this.repository.updateStatus(id, OrderStatus.CANCELLED);

    // Publish order.cancelled event (compensating transaction trigger)
    if (updatedOrder) {
      try {
        const event = createOrderCancelledEvent(updatedOrder);
        await this.kafkaProducer.publish('order.cancelled', id, event);
        console.log(`[OrderService] Compensating transaction - Published order.cancelled for order: ${id}`);
      } catch (error) {
        console.error(`[OrderService] Failed to publish order.cancelled event:`, error);
      }
    }

    return updatedOrder;
  }

  /**
   * Mark order as cancelled due to insufficient stock.
   * This is Step 3 of the Saga (failure path).
   */
  async markInsufficientStock(id: string): Promise<Order | null> {
    const order = await this.repository.findById(id);
    if (!order) return null;

    if (!isValidTransition(order.status, OrderStatus.CANCELLED)) {
      console.warn(`[OrderService] Invalid transition: ${order.status} → CANCELLED for order ${id}`);
      return order;
    }

    const updatedOrder = await this.repository.updateStatus(id, OrderStatus.CANCELLED);
    console.log(`[OrderService] Order ${id} cancelled - insufficient stock`);

    return updatedOrder;
  }
}

File: python-soa/order-service/app/application/order_service.py
"""
Order Service Implementation
===============================
Implements the order creation saga orchestration.

Design Pattern: Saga (Orchestration)
======================================
The Saga pattern manages distributed transactions across services
without using traditional 2-phase commit (which doesn't scale in SOA).

Saga Flow:
1. Client creates order → status = PENDING
2. Order Service publishes 'order.created' event
3. Inventory Service attempts to reserve stock
4. If stock available → publishes 'stock.reserved'
   - Order Service confirms order → status = CONFIRMED
5. If stock insufficient → publishes 'stock.insufficient'
   - Order Service cancels order → status = CANCELLED (compensation)

Compensation:
- If a confirmed order is cancelled, 'order.cancelled' is published
- Inventory Service releases the reserved stock

This ensures eventual consistency across the distributed system.
"""

import logging
from decimal import Decimal
from uuid import UUID, uuid4

from app.domain.model import Order, OrderItem, OrderStatus
from app.infrastructure.persistence.order_repository import OrderRepository
from app.infrastructure.messaging.kafka_producer import KafkaEventPublisher
from app.infrastructure.messaging.events import OrderEvents

logger = logging.getLogger(__name__)


class OrderServiceImpl:
    """
    Order Service implementation with Saga orchestration.

    Acts as the saga orchestrator, coordinating the distributed
    transaction between Order and Inventory services.
    """

    def __init__(self, repository: OrderRepository, publisher: KafkaEventPublisher):
        self._repository = repository
        self._publisher = publisher

    async def create_order(
        self,
        user_id: UUID,
        items: list[dict],
    ) -> Order:
        """
        Create a new order and initiate the order creation saga.

        Steps:
        1. Create Order aggregate with items
        2. Persist to database with PENDING status
        3. Publish 'order.created' event to trigger inventory reservation

        The saga continues asynchronously when the Inventory Service
        responds with either 'stock.reserved' or 'stock.insufficient'.
        """
        order_id = uuid4()

        # Build order items
        order_items = []
        for item_data in items:
            unit_price = Decimal(str(item_data["unit_price"]))
            quantity = item_data["quantity"]
            subtotal = unit_price * quantity

            order_item = OrderItem(
                id=uuid4(),
                order_id=order_id,
                product_id=UUID(item_data["product_id"]),
                product_name=item_data.get("product_name", ""),
                quantity=quantity,
                unit_price=unit_price,
                subtotal=subtotal,
            )
            order_items.append(order_item)

        # Create order aggregate
        order = Order(
            id=order_id,
            user_id=user_id,
            status=OrderStatus.PENDING,
            items=order_items,
        )
        order.calculate_total()

        # Persist order
        await self._repository.save(order)
        logger.info(
            "Order created: %s (user: %s, total: %s, status: PENDING)",
            order.id, order.user_id, order.total_amount,
        )

        # Publish event to initiate saga (Inventory Service will consume this)
        event = OrderEvents.order_created(order)
        await self._publisher.publish("order.created", str(order.id), event)
        logger.info("Published order.created event - saga initiated for order %s", order.id)

        return order

    async def get_by_id(self, order_id: UUID) -> Order | None:
        """Retrieve an order with its items."""
        return await self._repository.find_by_id(order_id)

    async def get_by_user(self, user_id: UUID) -> list[Order]:
        """Retrieve all orders for a user."""
        return await self._repository.find_by_user(user_id)

    async def confirm_order(self, order_id: UUID) -> None:
        """
        Saga callback: Confirm order after successful stock reservation.

        Called by the InventoryEventConsumer when 'stock.reserved' is received.
        Transitions order from PENDING → CONFIRMED.
        Publishes 'order.confirmed' event.
        """
        order = await self._repository.find_by_id(order_id)
        if not order:
            logger.error("Cannot confirm order %s: not found", order_id)
            return

        try:
            order.confirm()
            await self._repository.update_status(order_id, order.status)
            logger.info("Order confirmed: %s (saga completed successfully)", order_id)

            # Publish confirmation event
            event = OrderEvents.order_confirmed(order)
            await self._publisher.publish("order.confirmed", str(order_id), event)

        except ValueError as e:
            logger.warning("Cannot confirm order %s: %s", order_id, e)

    async def cancel_order(self, order_id: UUID, reason: str = "") -> None:
        """
        Saga callback/API: Cancel an order.

        Called when:
        - 'stock.insufficient' event received (saga compensation)
        - User manually cancels via API

        If the order was CONFIRMED, publishes 'order.cancelled' to trigger
        stock release in the Inventory Service (compensation).
        """
        order = await self._repository.find_by_id(order_id)
        if not order:
            logger.error("Cannot cancel order %s: not found", order_id)
            return

        was_confirmed = order.status == OrderStatus.CONFIRMED

        try:
            order.cancel()
            await self._repository.update_status(order_id, order.status)
            logger.info("Order cancelled: %s (reason: %s)", order_id, reason)

            # If order was confirmed, we need to release reserved stock
            if was_confirmed:
                event = OrderEvents.order_cancelled(order, reason)
                await self._publisher.publish("order.cancelled", str(order_id), event)
                logger.info(
                    "Published order.cancelled event for stock release (order: %s)",
                    order_id,
                )

        except ValueError as e:
            logger.warning("Cannot cancel order %s: %s", order_id, e)
